### ISSUE TITLE
Question Marks break Exact Test

### ISSUE THREAD

**Issue by sparry0407** [jcrodriguez-dis/moodle-mod_vpl#117]
If exact text in a test case contains a question mark, it breaks the comparison. so, if you have a test of:
`output="ABC?"`
This will fail a program that outputs:
`ABC?`

**Comment by jcrodriguez-dis** [comment]
I have not been able to reproduce this issue.
Please, show your vpl_evaluate.cases file and the student's program that does not pass the test.

Notice that your case definition does not accept anything after the "?" neither a newline.
To define a case that accepts exact outputs with and without a newline, you can use two outputs in this test case
```
output="ABC?"
output="ABC?
"
```

If some output defined matches the program output, the test case is passed.

**Comment by sparry0407** [comment]
This one works:
[cid:0dd0a328-e4a2-4888-a525-4cafd820ede1]
[cid:f2c86be4-1f3d-42c3-9937-b288b4d770e2][cid:15197f69-9cb1-48dd-90e2-ff377cfd5b35]
This one does not:
[cid:51c08b53-7b01-4551-886d-d4ea9693571b]
[cid:ffb2cbef-bca9-44a5-823b-3a530ffa5f8a][cid:97353f6f-5dbc-4240-9818-fe0c43e7bf76]

________________________________

**Comment by jcrodriguez-dis** [comment]
Sorry, I can not see your images.

I think that may be better to use the GitHub interface than reply by email.

[You can use this link to access the thread](https://github.com/jcrodriguez-dis/moodle-mod_vpl/issues/117)

**Comment by sparry0407** [comment]
Correct:
![001](https://user-images.githubusercontent.com/67392827/97701792-e347ed80-1aa5-11eb-9524-e45df92c46c5.png)
![002](https://user-images.githubusercontent.com/67392827/97701808-e93dce80-1aa5-11eb-8d2b-c4e053976447.png)
![003](https://user-images.githubusercontent.com/67392827/97701816-eba02880-1aa5-11eb-94f4-097ad1e3edb4.png)

Incorrect:
![004](https://user-images.githubusercontent.com/67392827/97701822-ee028280-1aa5-11eb-9652-7970aaab30f4.png)
![005](https://user-images.githubusercontent.com/67392827/97701850-f6f35400-1aa5-11eb-819f-17cad6e3a4e6.png)
![006](https://user-images.githubusercontent.com/67392827/97701864-f9ee4480-1aa5-11eb-88d9-7d3c8e0ff9ea.png)

**Comment by sparry0407** [comment]
I tried your suggestion, which worked. It does seem there is an inconsistency between how the tests are treated depending on whether question marks (and possibly other symbols) are in the text or not.

**Comment by jcrodriguez-dis** [comment]
Yes, you are right.

I have reviewed the code and found that the exact text matcher trims (cleans spaces, tabs, and newlines) the right side of the output if the searched text end with an alpha char.
This code was introduced to avoid frequent fails by newlines at the end of the output.
This code does that the "ABC" matches with print("ABC") in python, but really print("ABC") print also a newline.

I tried to reproduce the problem using C, but without including a "\n" at the end of the printf() this is why I don't found the problem.

Thanks for your report, I will fix this bug in the next release.
